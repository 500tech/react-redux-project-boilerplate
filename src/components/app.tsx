import * as React from 'react';
import { ThemeProvider } from 'emotion-theming';
import { Router } from 'react-router-dom';
import { Route } from 'react-router';
import { Provider } from 'react-redux';

import history from 'utils/history.utils';
// TODO: remove if no need for Lazy load routes:
import lazyLoad from 'utils/lazy-load.utils';

import store from 'store';
import theme from 'constants/themes.constants';

import Localization from 'components/localization'; // TODO: remove if no localization
import Layout from 'components/layout/layout';
import Sample from 'sample/sample'; // TODO: replace this with actual component

class App extends React.Component<{}> {
  render() {
    return (
      <Provider store={store}>
        <Localization>
          <ThemeProvider theme={theme}>
            <Router history={history}>
              <Layout>
                <Route exact path="/" name="sample" component={Sample} />
                <Route
                  path="/lazy"
                  name="lazy"
                  component={lazyLoad(() => import('sample/lazy'))}
                />
              </Layout>
            </Router>
          </ThemeProvider>
        </Localization>
      </Provider>
    );
  }
}

export default App;
