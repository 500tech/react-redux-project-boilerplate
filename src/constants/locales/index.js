// @flow
import enUS from 'constants/locales/en-US';

export default {
  'en-US': enUS
};
