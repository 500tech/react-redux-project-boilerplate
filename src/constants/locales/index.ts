import enUS from 'constants/locales/en-US';
import heIL from 'constants/locales/he-IL';
import { Locale } from 'types/common.types';

export type LocaleTypes = 'en-US' | 'he-IL';

interface LocaleList {
  [key: string]: Locale;
}

const locales: LocaleList = {
  'en-US': enUS,
  'he-IL': heIL
};

export default locales;
