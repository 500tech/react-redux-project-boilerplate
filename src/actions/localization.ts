// TODO: remove this file if no localization is needed
import { BaseAction } from 'types/base-redux';
import { LocaleTypes } from 'constants/locales';

export const SET_LOCALE: string = '[localization] Set Locale';

export interface SetLocaleAction extends BaseAction {
  payload: {
    locale: string;
  };
}

export const setLocale = (locale: LocaleTypes) => ({
  type: SET_LOCALE,
  payload: {
    locale
  }
});
