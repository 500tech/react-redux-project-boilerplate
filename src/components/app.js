// @flow
import * as React from 'react';
import { ThemeProvider } from 'styled-components';
import { Router } from 'react-router-dom';
import { Route } from 'react-router';
import { Provider } from 'react-redux';
// TODO: remove if no need for Lazy load routes:
import Loadable from 'react-loadable';

import history from 'utils/history.utils';
import store from 'store';
import * as themes from 'constants/themes.constants';

import Localization from 'components/localization'; // TODO: remove if no localization
import Layout from 'components/layout/layout';

import Sample from 'sample/sample'; // TODO: replace this with actual component
import Loading from 'sample/loading';

// TODO: remove if no need for Lazy load routes:
const lazyLoad = loader =>
  Loadable({
    loader,
    loading: () => <Loading showLoading={true} />
  });

class App extends React.Component<{||}> {
  render() {
    return (
      <Provider store={store}>
        <Localization>
          <ThemeProvider theme={themes.main}>
            <Router history={history}>
              <Layout>
                <Route exact path="/" name="sample" component={Sample} />
                <Route
                  path="/lazy"
                  name="lazy"
                  component={lazyLoad(() => import('sample/lazy'))}
                />
              </Layout>
            </Router>
          </ThemeProvider>
        </Localization>
      </Provider>
    );
  }
}

export default App;
