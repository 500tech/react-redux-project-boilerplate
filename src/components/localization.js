// @flow
import * as React from 'react';
import forEach from 'lodash/forEach';
import { connect } from 'react-redux';
import { IntlProvider, addLocaleData } from 'react-intl';
import moment from 'moment';

import store from 'store';

import locales from 'constants/locales';

import type { State } from 'types/redux.types';
import type { MapStateToProps } from 'react-redux';

forEach(locales, (locale, key) =>
  moment.defineLocale(key, locale.dateTimeFormat)
);

// Go over all of the available locales and register them
forEach(locales, (value, key) => {
  addLocaleData({
    locale: key,
    // Couldn't find any documentation about 'pluralRuleFunction', throws error if not present
    pluralRuleFunction: () => {}
  });
});

type ConnectedProps = {
  locale: string
};

type OwnProps = {
  children: React.Node
};

export const Localization = ({
  locale,
  children
}: ConnectedProps & OwnProps) => (
  <IntlProvider
    locale={locale}
    key={locale}
    messages={locales[locale].translations}>
    {children}
  </IntlProvider>
);

const mapStateToProps: MapStateToProps<State, OwnProps, {}> = ({
  localization
}: State) => ({
  locale: localization.locale
});

let currentLocale = store.getState().localization.locale;

moment.locale(currentLocale);
store.subscribe(() => {
  const newLocale = store.getState().localization.locale;

  if (newLocale !== currentLocale) {
    currentLocale = newLocale;
    moment.locale(currentLocale);
  }
});

export default connect(mapStateToProps, {})(Localization);
