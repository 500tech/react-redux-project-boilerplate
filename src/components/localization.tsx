import * as React from 'react';
import forEach from 'lodash/forEach';
import { connect } from 'react-redux';
import { IntlProvider } from 'react-intl';
import moment, { LocaleSpecification } from 'moment';

import store from 'store';

import locales, { LocaleTypes } from 'constants/locales';

import { State } from 'types/redux.types';

forEach(locales, (locale, key) =>
  moment.defineLocale(
    key,
    (locale.dateTimeFormat as unknown) as LocaleSpecification
  )
);

if (!Intl.PluralRules) {
  require('@formatjs/intl-pluralrules/polyfill');
}

// @ts-ignore
if (!Intl.RelativeTimeFormat) {
  require('@formatjs/intl-relativetimeformat/polyfill');
}

// Go over all of the available locales and register them
forEach(locales, (_value, key: string = '') => {
  const localeKey = key.split('-')[0];

  if (!localeKey) {
    return;
  }

  require(`@formatjs/intl-pluralrules/dist/locale-data/${localeKey}`);
  require(`@formatjs/intl-relativetimeformat/dist/locale-data/${localeKey}`);
});

export const Localization = ({ locale, children }: Props) => (
  <IntlProvider
    locale={locale}
    key={locale}
    messages={locales[locale].translations}>
    {children}
  </IntlProvider>
);

interface OwnProps {
  children: React.ReactNode;
}

interface StateProps {
  locale: LocaleTypes;
}

type Props = StateProps & OwnProps;

const mapStateToProps = (state: State): StateProps => ({
  locale: state.localization.locale
});

let currentLocale = (store.getState() as State).localization.locale;

moment.locale(currentLocale);
store.subscribe(() => {
  const newLocale = (store.getState() as State).localization.locale;

  if (newLocale !== currentLocale) {
    currentLocale = newLocale;
    moment.locale(currentLocale);
  }
});

export default connect(mapStateToProps)(Localization);
