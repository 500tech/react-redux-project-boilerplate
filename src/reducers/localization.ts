// TODO: remove this file if no localization is needed
import { get, set } from 'lodash/fp';
import { handleActions } from 'redux-actions';

import { setLocale } from 'actions/localization';

import * as AT from 'actions/localization';
import { LocaleTypes } from 'constants/locales';

export interface LocalizationState {
  locale: LocaleTypes;
}

const initialState: LocalizationState = {
  locale: 'en-US'
};

const localization = handleActions<LocalizationState>(
  {
    [AT.SET_LOCALE]: (
      state: LocalizationState,
      action: ReturnType<typeof setLocale>
    ): LocalizationState => {
      const locale = get('payload.locale', action);

      return set('locale', locale, state);
    }
  },
  initialState
);

export default localization;
