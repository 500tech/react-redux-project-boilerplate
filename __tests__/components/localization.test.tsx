import React from 'react';
import { shallow } from 'enzyme';
import { Localization } from '../../src/components/localization';
import { LocaleTypes } from '../../src/constants/locales';

const setup = ({ locale = 'en-US' as LocaleTypes } = {}) => ({
  locale
});

describe('<Localization />', () => {
  test('should render', () => {
    const props = setup();
    const component = shallow(
      <Localization {...props}>
        <div>Hello world</div>
      </Localization>
    );

    expect(component).toMatchSnapshot();
  });
});
