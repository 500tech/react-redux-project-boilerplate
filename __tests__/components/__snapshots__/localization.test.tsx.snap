// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Localization /> should render 1`] = `
<IntlProvider
  key="en-US"
  locale="en-US"
  messages={
    Object {
      "sample.common.loading": "Loading...",
      "sample.homepage.description": "This is a sample component that uses a sample action + reducer, it fetches posts from a remote server and displays them",
      "sample.homepage.title": "Welcome to the 500Tech boilerplate",
    }
  }
>
  <div>
    Hello world
  </div>
</IntlProvider>
`;
